File: Python/Stage 3/S3_PIPE_02.py
def compose(*fns):
    def apply(value):
        result = value
        for fn in reversed(fns):
            result = fn(result)
        return result

    return apply


add2 = lambda n: n + 2
mul3 = lambda n: n * 3
sub1 = lambda n: n - 1

piped = lambda x: sub1(mul3(add2(x)))
composed = compose(sub1, mul3, add2)

print(piped(4))
print(composed(4))

File: Js/Task 5/S3_PIPE_02.js
function compose(...fns) {
    return x => fns.reduceRight((v, f) => f(v), x);
}

const add = x => x + 1;
const multiply = x => x * 2;

const result = compose(multiply, add)(3);  // multiply(add(3)) = (3 + 1) * 2 = 8
console.log(result);